src/transport/serializer.h

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+#pragma once
+
+#include "serialized-data.h"
+
+#include <napa/module/binding/wraps.h>
+
+namespace napa {
+namespace transport {
+
+    using namespace v8;
+
+    typedef std::pair<SharedArrayBuffer::Contents, std::shared_ptr<ExternalizedContents>> SharedArrayBufferExternalization;
+
+    class Serializer : public ValueSerializer::Delegate {
+    public:
+        explicit Serializer(Isolate* isolate) :
+            _isolate(isolate),
+            _serializer(isolate, this) {}
+
+        Maybe<bool> WriteValue(Local<Value> value);
+
+        std::shared_ptr<SerializedData> Release() { return _data; }
+
+    protected:
+        void ThrowDataCloneError(Local<String> message) override;
+
+        Maybe<uint32_t> GetSharedArrayBufferId(
+            Isolate* isolate,
+            Local<SharedArrayBuffer> shared_array_buffer
+        ) override;
+
+        void* ReallocateBufferMemory(void* old_buffer, size_t size, size_t* actual_size) override;
+
+        void FreeBufferMemory(void* buffer) override;
+
+    private:
+        SharedArrayBufferExternalization MaybeExternalize(Local<SharedArrayBuffer> shared_array_buffer);
+
+        Maybe<bool> FinalizeTransfer();
+
+        Isolate* _isolate;
+        ValueSerializer _serializer;
+        std::shared_ptr<SerializedData> _data;
+        std::vector<Global<SharedArrayBuffer>> _shared_array_buffers;
```

Comment:
What's this for?

Reply:
You should mean this one mentioned this afternoon.
1) this is used by GetSharedArrayBufferId. It generates a ID for SAB and accumulate SAB instances encountered during parsing the objects, which might be composite object containing multiple and potentially duplicated SharedArrayBuffer instances. 
2) Why it's a vector? We suppose there will not a large number of  SAB instances for a passed in objects. It easy to be manipulated in GetSharedArrayBufferId and FinalizeTransfer.   It might be practically better in terms of performance. 
3) Lifecycle: This vector is held by serializer. Serializer instance is a local variable created in TransportUtils::SerializeValue(...). It will be disposed out of TransportUtils::SerializeValue
maybe GLobal<SAB> is confusing. It copied from d8 implementation. Let me try to replace it by Shared_Ptr

---
In reply to: [155907881](https://github.com/Microsoft/napajs/pull/137#discussion_r155907881) [](ancestors = 155907881)